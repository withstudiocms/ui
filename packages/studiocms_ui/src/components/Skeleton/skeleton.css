@keyframes sui-skeleton-slide {
	0% {
		background-position: 175% 0;
	}
	100% {
		background-position: -75% 0;
	}
}

@keyframes sui-skeleton-pulse {
	0% {
    background-color: hsl(var(--background-step-2));
  }
  50% {
    background-color: hsl(var(--background-step-3));
  }
  100% {
    background-color: hsl(var(--background-step-2));
  }
}

.sui-skeleton {
	display: inline-block;
	position: relative;
	width: 100%;
	height: 100%;
	overflow: hidden;
	background-color: hsl(var(--background-step-2));
}

.sui-skeleton-card {
  display: flex;
  flex-direction: column;
  background-color: hsl(var(--background-step-1));
  padding: 1rem;
  gap: 0.5rem;
  height: 100%;
  overflow: hidden;
}

.sui-skeleton-card.sui-skeleton-row {
  flex-direction: row;
}

.sui-skeleton-justify-start, .sui-skeleton-align-start {
  justify-content: flex-start;
}

.sui-skeleton-justify-center, .sui-skeleton-align-center {
  justify-content: center;
}

.sui-skeleton-justify-end, .sui-skeleton-align-end {
  justify-content: flex-end;
}

.sui-skeleton-items-start, .sui-skeleton-justify-start {
  align-items: flex-start;
}

.sui-skeleton-items-center, .sui-skeleton-justify-center {
  align-items: center;
}

.sui-skeleton-items-end, .sui-skeleton-justify-end {
  align-items: flex-end;
}

.sui-skeleton-text {
  display: inline-block;
  background-color: hsl(var(--background-step-2));
  height: 1rem;
}

.sui-skeleton-block {
  display: block;
  background-color: hsl(var(--background-step-2));
  height: 1rem;
}

.sui-skeleton-circle {
  display: inline-block;
  background-color: hsl(var(--background-step-2));
  width: 100%;
  height: 100%;
  border-radius: var(--radius-full);
}

.sui-skeleton-radii-sm {
  border-radius: var(--radius-sm);
}

.sui-skeleton-radii-md {
  border-radius: var(--radius-md);
}

.sui-skeleton-radii-lg {
  border-radius: var(--radius-lg);
}

.sui-skeleton-anim-none {
  animation: none;
}

.sui-skeleton-anim-slide {
  @media (prefers-reduced-motion: no-preference) {
    background-image: linear-gradient(90deg, hsl(var(--background-step-2)) 25%, hsl(var(--background-step-3)) 50%, hsl(var(--background-step-2)) 75%);
    background-position: 175% 0;
    background-size: 200% 100%;
    background-repeat: no-repeat;
    animation: sui-skeleton-slide 2s ease-in-out infinite;
  }
}

.sui-skeleton-anim-pulse {
  @media (prefers-reduced-motion: no-preference) {
    animation: sui-skeleton-pulse 2s linear infinite;
  }
}
