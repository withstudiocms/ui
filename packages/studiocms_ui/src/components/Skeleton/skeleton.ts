import { z } from 'astro/zod';

const measurementPx = z.custom<`${number}px`>();
const measurementRem = z.custom<`${number}rem`>();
const measurementPercent = z.custom<`${number}%`>();

const measurement = z.union([measurementPx, measurementRem, measurementPercent]);

const BaseSkeletonSchema = z.object({
	/**
	 * The variant of the skeleton.
	 */
	variant: z.literal('text'),

  /**
   * The class of the skeleton. Defaults to `undefined`.
   */
	class: z.string().optional(),

  /**
   * The radius of the skeleton. Defaults to `lg`.
   */
	radius: z.union([z.literal('none'), z.literal('sm'), z.literal('md'), z.literal('lg')]).optional(),

  /**
   * The animation of the skeleton. Defaults to `slide`.
   */
	animation: z.union([z.literal('none'), z.literal('slide'), z.literal('pulse')]).optional(),

  /**
   * The width of the skeleton. Defaults to `100%`.
   */
	width: measurement.optional(),

  /**
   * The height of the skeleton. Defaults to `100%`.
   */
	height: measurement.optional(),
});

const CardSkeletonSchema = BaseSkeletonSchema.extend({
  /**
   * The variant of the skeleton. `card` is used for cards which contain other Skeleton components.
   */
	variant: z.literal('card'),

  /**
   * The direction of the skeleton. Defaults to `column`.
   */
	direction: z.union([z.literal('row'), z.literal('column')]).optional(),

  /**
   * The horizontal alignment of the skeleton. Defaults to `center`.
   */
	hAlign: z.union([z.literal('start'), z.literal('center'), z.literal('end')]).optional(),

  /**
   * The vertical alignment of the skeleton. Defaults to `center`.
   */
	vAlign: z.union([z.literal('start'), z.literal('center'), z.literal('end')]).optional(),

  /**
   * The gap between the skeletons. Defaults to `0.5rem`.
   */
	gap: measurement.optional(),
});

const BlockCircleSchema = BaseSkeletonSchema.extend({
  /**
   * The variant of the skeleton. `circle` is used for circles which contain other Skeleton components.
   */
	variant: z.union([z.literal('circle'), z.literal('block')]),

  /**
   * The width of the skeleton.
   */
	width: measurement,

  /**
   * The height of the skeleton.
   */
	height: measurement,
});

export const SkeletonSchema = z.union([BaseSkeletonSchema, CardSkeletonSchema, BlockCircleSchema]);