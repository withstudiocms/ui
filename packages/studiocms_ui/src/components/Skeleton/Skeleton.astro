---
import './skeleton.css';

interface BaseSkeletonProps {
	/**
	 * The class of the skeleton. Defaults to `undefined`.
	 */
	class?: string;

	/**
	 * The radius of the skeleton. Defaults to `lg`.
	 * 
	 * This property is ignored for the `circle` variant.
	 */
	radius?: 'none' | 'sm' | 'md' | 'lg';

	/**
	 * The animation of the skeleton. Defaults to `slide`.
	 * 
	 * This property defaults to none when the user has reduced motion enabled.
	 */
	animation?: 'none' | 'slide' | 'pulse';
}

interface TextSkeletonProps extends BaseSkeletonProps {
	/**
	 * The variant of the skeleton. `text` is used for long lines of text.
	 */
	variant: 'text';

	/**
	 * The width of the skeleton. Defaults to `100%`.
	 */
	width?: `${number}%` | `${number}px` | `${number}rem`;

	/**
	 * The height of the skeleton. Defaults to `100%`.
	 */
	height?: `${number}%` | `${number}px` | `${number}rem`;
}

interface CardSkeletonProps extends BaseSkeletonProps {
	/**
	 * The variant of the skeleton. `card` is used for cards which contain other Skeleton components.
	 */
	variant: 'card';

	/**
	 * The width of the skeleton. Defaults to `100%`.
	 */
	width?: `${number}%` | `${number}px` | `${number}rem`;

	/**
	 * The height of the skeleton. Defaults to `100%`.
	 */
	height?: `${number}%` | `${number}px` | `${number}rem`;

	/**
	 * The direction of the skeleton. Defaults to `column`.
	 */
	direction?: 'row' | 'column';

	/**
	 * The horizontal alignment of the skeleton. Defaults to `center`.
	 */
	hAlign?: 'start' | 'center' | 'end';

	/**
	 * The vertical alignment of the skeleton. Defaults to `center`.
	 */
	vAlign?: 'start' | 'center' | 'end';

	/**
	 * The gap between the skeletons. Defaults to `0.5rem`.
	 */
	gap?: `${number}px` | `${number}rem`;
}

interface BlockSkeletonProps extends BaseSkeletonProps {
	/**
	 * The variant of the skeleton. `block` is used for parallelograms.
	 */
	variant: 'block';

	/**
	 * The width of the skeleton.
	 */
	width: `${number}%` | `${number}px` | `${number}rem`;

	/**
	 * The height of the skeleton.
	 */
	height: `${number}%` | `${number}px` | `${number}rem`;
}

interface CircleSkeletonProps extends BaseSkeletonProps {
	/**
	 * The variant of the skeleton. `circle` is used for circular elements.
	 */
	variant: 'circle';

	/**
	 * The width of the skeleton.
	 */
	width: `${number}%` | `${number}px` | `${number}rem`;

	/**
	 * The height of the skeleton.
	 */
	height: `${number}%` | `${number}px` | `${number}rem`;
}

type Props = TextSkeletonProps | CardSkeletonProps | BlockSkeletonProps | CircleSkeletonProps;

const { 
	variant,
	width,
	height,
	radius = 'lg',
	animation = 'slide',
	class: className,
} = Astro.props;

const gap = variant === 'card' ? Astro.props.gap ?? '0.5rem' : undefined;

const styles = [
 width && `width: ${width};`,
 height && `height: ${height};`,
 gap && `gap: ${gap};`
].filter(Boolean).join(' ');

function getClasses() {
	const classes = ['sui-skeleton'];

	if (variant) {
		classes.push(`sui-skeleton-${variant}`);
		
		if (variant === 'card') {
			const direction = Astro.props.direction ?? 'column';
			const hAlign = Astro.props.hAlign ?? 'center';
			const vAlign = Astro.props.vAlign ?? 'center';
			
			classes.push(`sui-skeleton-${direction}`);
			
			const hAlignProp = direction === 'row' ? 'justify' : 'align';
			const vAlignProp = direction === 'row' ? 'items' : 'justify';
			
			classes.push(`sui-skeleton-${hAlignProp}-${hAlign}`);
			classes.push(`sui-skeleton-${vAlignProp}-${vAlign}`);
		}
	}

	if (radius !== 'none' && variant !== 'circle') {
		classes.push(`sui-skeleton-radii-${radius}`);
	}

	if (animation !== 'none' && variant !== 'card') {
		classes.push(`sui-skeleton-anim-${animation}`);
	}
	if (className) {
		classes.push(className);
	}

	return classes;
}
---

<div class:list={getClasses()} style={styles}>
	<slot />
</div>