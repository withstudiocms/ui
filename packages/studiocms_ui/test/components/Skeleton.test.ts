import { describe, expect } from 'vitest';
import Skeleton from '../../src/components/Skeleton/Skeleton.astro';
import { test } from '../fixtures/vitest/AstroContainer';

describe('Skeleton Component', () => {
	test('renders Skeleton component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Skeleton, 'Skeleton', { props: { variant: 'card' } });
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ variant: 'text' },
		{ variant: 'circle', width: '50%', height: '8rem' },
		{ variant: 'block', width: '100px', height: '100px' },
	])('renders Skeleton component with variant: $variant', async (props, { renderComponent }) => {
		const result = await renderComponent(Skeleton, 'Skeleton', {
			props,
		});
		expect(result).toMatchSnapshot();
	});
});
