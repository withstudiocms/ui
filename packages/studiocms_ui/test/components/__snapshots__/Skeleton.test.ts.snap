// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Skeleton Component > renders Skeleton component correctly 1`] = `"<div class="sui-skeleton sui-skeleton-card sui-skeleton-column sui-skeleton-align-center sui-skeleton-justify-center sui-skeleton-radii-lg" style="gap: 0.5rem;">  </div>"`;

exports[`Skeleton Component > renders Skeleton component with variant: 'block' 1`] = `"<div class="sui-skeleton sui-skeleton-block sui-skeleton-radii-lg sui-skeleton-anim-slide" style="width: 100px; height: 100px;">  </div>"`;

exports[`Skeleton Component > renders Skeleton component with variant: 'circle' 1`] = `"<div class="sui-skeleton sui-skeleton-circle sui-skeleton-anim-slide" style="width: 50%; height: 8rem;">  </div>"`;

exports[`Skeleton Component > renders Skeleton component with variant: 'text' 1`] = `"<div class="sui-skeleton sui-skeleton-text sui-skeleton-radii-lg sui-skeleton-anim-slide" style>  </div>"`;
